pyiron_continuum/schroedinger/mesh.py

```diff
@@ -0,0 +1,190 @@
+        laplacian: Given a callable that takes the mesh as its argument, or a numpy array with the same shape as the
+            mesh's real-space dimensions (i.e. `self.shape[1:]`, since the first mesh dimension maps over the dimensions
+            themselves) returns the discrete Laplace operator on this mesh applied to that funcation/data.
+    """
+
+    def __init__(self, bounds=1, divisions=1, simplify_1d=False):
+        """
+        Instantiate a rectangular mesh.
+
+        Args:
+            bounds (float/list/numpy.ndarray): The upper and lower bounds for each dimension of the mesh. A single
+                value, L, creates a mesh on [0, L]. A list/array with shape (n<=3,) creates an n-dimensional mesh with
+                on [0, L_i] for each of the L_i values given. A two-dimensional list/array should have shape (n<=3,2)
+                and gives lower and upper bounds for each dimension, i.e. `[[0.5, 1]]` makes a 1D mesh on [0.5, 1].
+            divisions (int/list/numpy.ndarray): How many grid divisions to use in each dimension. An integer will be
+                mapped up to give that number of divisions on all dimensions provided in `bounds`, otherwise the
+                dimensionality between the two arguments must agree.
+            simplify_1d (bool): Whether to simplify the output for 1D meshes so they have shape (n,) instead of (1, n).
+                (Default is False.)
+        """
+        super().__init__()
+        bounds, divisions = self._clean_input(bounds, divisions)
+        self.storage.bounds = bounds
+        self.storage.divisions = divisions
+        self.storage.simplify_1d = simplify_1d
+        self._build_mesh()
+
+    @property
+    def bounds(self):
+        return self.storage.bounds
+
+    @bounds.setter
+    def bounds(self, new_bounds):
+        new_bounds, _ = self._clean_input(new_bounds, self.divisions)
+        self.storage.bounds = new_bounds
+        self._build_mesh()
+
+    @property
+    def divisions(self):
+        return self.storage.divisions
+
+    @divisions.setter
+    def divisions(self, new_divisions):
+        _, new_divisions = self._clean_input(self.bounds, new_divisions)
+        self.storage.divisions = new_divisions
+        self._build_mesh()
+
+    @property
+    def simplify_1d(self):
+        return self.storage.simplify_1d
+
+    @simplify_1d.setter
+    def simplify_1d(self, simplify: bool):
+        self.storage.simplify_1d = simplify
+
+    def _simplify_1d(self, x):
+        if len(x) == 1 and self.storage.simplify_1d:
+            return np.squeeze(x)
+        else:
+            return x
+
+    @property
+    def mesh(self) -> np.ndarray:
+        return self._simplify_1d(self.storage.mesh)
+
+    @property
+    def steps(self) -> np.ndarray:
+        """Spacing between each mesh point."""
+        return self._simplify_1d(self.storage.steps)
+
+    @property
+    def shape(self):
+        return self.mesh.shape
+
+    @property
+    def lengths(self):
```

Comment:
Is there a particular reason that `bounds` and `lengths` exist separately? They return the same thing, right?

Reply:
Only if bounds start at zero, which is default but not necessary

Reply:
Ah! Then I have a different concern: In line 144 it assumes that a 1d-array defines the lengths, but it's not straightforwardly clear to me why the user shouldn't have wanted a 1d mesh with those bounds. For example, `bounds = [2, 5]` would currently be translated to `bounds = [[0, 2], [0, 5]]`, but the user might have actually wanted `bounds = [[2, 5]]`.

Reply:
And the definition of `lengths` can also be simplified to `return np.diff(self.bounds, axis=-1).flatten()`.

Reply:
> Ah! Then I have a different concern: In line 144 it assumes that a 1d-array defines the lengths, but it's not straightforwardly clear to me why the user shouldn't have wanted a 1d mesh with those bounds. For example, `bounds = [2, 5]` would currently be translated to `bounds = [[0, 2], [0, 5]]`, but the user might have actually wanted `bounds = [[2, 5]]`.

Please read through my earlier discussion with Raynol here

Reply:
Ah here we go, at least both @raynol-dsouza and I had the same problem, so I guess we'll see more people having the same problem in the future.

Actually I don't really understand the advantage of the non-zero origin. Do you have a specific case in mind?

Reply:
If you don't want nonzero start points then there's nothing to complain about, just enter a one, two, or three-length list of the endpoints you want. If you *do* for some reason want nonzero startpoints *and* you're only doing 1d, then, yes, you need to remember to add the extra set of brackets. I really don't think this is such a burden.

As for use case, sure, imagine you want to use the mesh to evaluate a function that's divergent at zero.

Reply:
fine then it should at least make sure that there's no 0 in `bounds` if it's a 1d array, because I can imagine people doing something like `bounds = [0, 10]` while trying to create a 1d-mesh of length 10.

Reply:
Yep, good call. A check that no dimension is length zero is wise, and [0,X] is probably the most likely cause